shared/fs/header.js

```diff
@@ -0,0 +1,167 @@
+  fontSize: 11,
+  marginLeft: globalMargins.xtiny,
+  marginRight: globalMargins.xtiny,
+}
+
+const styleTeamAvatar = {
+  marginRight: globalMargins.xtiny,
+}
+
+const FolderHeaderBreadcrumb = FolderHeaderBreadcrumbConnector(
+  ({visibility, item, onOpen}: FolderBreadcrumbConnectedProps) => {
+    return (
+      <ClickableBox onClick={onOpen}>
+        <Text type="BodySmallSemibold" style={headerPathStyleParent}>
+          {item}
+        </Text>
+      </ClickableBox>
+    )
+  }
+)
+
+const FolderHeaderItem = ({index, totalCrumbs, visibility, item, path}: FolderBreadcrumbProps) => {
+  let separator = <Icon type="iconfont-back" style={iconStyle} />
+  if (index === 0) {
+    separator = null
+  }
+  if (index === totalCrumbs - 1) {
+    return (
+      <Box style={folderBreadcrumbStyle}>
+        {separator}
+        {visibility === 'team' &&
+          index === 2 && <Avatar size={12} teamname={item} isTeam={true} style={styleTeamAvatar} />}
+        <Text type="BodyBig">{item}</Text>
+      </Box>
+    )
+  } else {
+    return (
+      <Box style={folderBreadcrumbStyle}>
+        {separator}
+        {visibility === 'team' &&
+          index === 2 && <Avatar size={12} teamname={item} isTeam={true} style={styleTeamAvatar} />}
+        <FolderHeaderBreadcrumb visibility={visibility} item={item} path={path} />
+      </Box>
+    )
+  }
+}
+
+class FolderHeader extends React.Component<FolderHeaderProps, FolderHeaderState> {
+  state: FolderHeaderState
+
+  constructor(props: FolderHeaderProps) {
+    super(props)
+    this.state = {
+      pathElems: Types.getPathElements(props.path),
+    }
+  }
+
+  componentWillReceiveProps = (nextProps: FolderHeaderProps) => {
+    this.setState(prevState => ({pathElems: Types.getPathElements(nextProps.path)}))
+  }
+
+  _createBreadcrumbs = () => {
+    switch (this.state.pathElems.length) {
+      case 1:
+        return rootHeaderContent
+      case 2:
+      // {private, public, team}
+      // fallthrough
+      case 3:
+        // TLF
+        const visibility = Types.getPathVisibility(this.props.path)
+        return (
+          <Box style={folderHeaderStyleTree}>
+            {this.state.pathElems.map((elem, index, pathElems) => {
+              // TODO: make this more efficient, currently it's doing repeated
+              // work for each item.
+              const pelems = pathElems.slice(0, index + 1)
+              const path = Types.stringToPath('/' + pelems.join('/'))
+              return (
+                <FolderHeaderItem
+                  key={index}
+                  index={index}
+                  visibility={visibility}
+                  item={elem}
+                  totalCrumbs={pathElems.length}
+                  path={path}
+                />
+              )
+            })}
+          </Box>
+        )
+      default:
+      // TODO: fix with list
+    }
+  }
+
+  render() {
+    return <Box>{this._createBreadcrumbs()}</Box>
+  }
+}
+
+export default FolderHeader
```

Comment:
overall i think this component is very large given the task at hand. If it's a big component i think there's value in splitting it up into sub components and maybe a class but i think this is pretty small. i didn't test this but this is what i'd do
```
// simple props
type Props = {
    items: Array<{
        name: string,
        pathTillItem: string,
    }>,
    isTeamPath: boolean,
    onClickItem: (pathTillItem: string) => void,
}

// style omitted for brevity
const Breadcrumbs = ({items, onClickItem, isTeamPath}: Props) => {
    const lastIndex = items.length - 1
    return <Box>
        {items.length === 1 
          ? <Box><Text type="BodyBig">Keybase Files</Text></Box> 
          : {items.map((i, idx) => {
                <ClickableBox onClick={() => onClickItem(item.pathTillTeam)}>
                    {i === 2 && isTeamPath && <Avatar size={12} teamname={item} isTeam={true} />}
                    <Text type="BodySmallSemibold">{item}</Text>
                    {i !== lastIndex && <Icon type="iconfont-back" style={iconStyle} />}
                </ClickableBox>
            })}
        }
    </Box>
}
```